File: backend/models/schemas.py
from collections import defaultdict
from enum import Enum
from typing import TypeAlias
from pydantic import BaseModel, ConfigDict, Field

# Graph data models

class NearestNodeAlgorithm(str, Enum):
    """Algorithm used by Graph.get_nearest_node()."""
    LINEAR = "linear"       # O(n) brute-force scan with euclidean distance
    KDTREE = "kdtree"       # scipy KD-tree (TODO)

class NodeData(BaseModel):
    """ 
    Geographic node with lat/lon. Extra OSM attributes allowed. 
    TODO: Add traffic scores (static, then dynamic as attrs)
    TODO: Tolls, croad quality.
    """
    lat: float
    lon: float
    model_config = ConfigDict(extra="allow")

class EdgeData(BaseModel):
    """Directed edge with distance and optional speed limit."""
    length_m: float
    max_speed_kmh: float | None = None
    model_config = ConfigDict(extra="allow")

# Graph-level type aliases for data structure

NodeMap: TypeAlias = dict[int, NodeData]
AdjacencyMap: TypeAlias = defaultdict[int, dict[int, EdgeData]]
NodeList: TypeAlias = list[tuple[float, float, int]]

# API request / response schemas

class LatLon(BaseModel):
    lat: float = Field(..., ge=-90, le=90)
    lon: float = Field(..., ge=-180, le=180)

class ShortestRequest(BaseModel):
    origin: LatLon
    destination: LatLon

class ShortestResponse(BaseModel):
    path: list[LatLon]
    distance_m: float

class MapNodesResponse(BaseModel):
    nodes: list[dict]  # [{"id", "lat", "lon"}, ...]

File: frontend/models/schemas.ts
import { z } from "zod";

// Zod schemas (mirror backend Pydantic models)
export const LatLonSchema = z.object({
  lat: z.number().min(-90).max(90),
  lon: z.number().min(-180).max(180),
});

export const ShortestResponseSchema = z.object({
  path: z.array(LatLonSchema),
  distance_m: z.number(),
});

export const MapNodesResponseSchema = z.object({
  nodes: z.array(
    z.object({
      id: z.number(),
      lat: z.number(),
      lon: z.number(),
    })
  ),
});

// Types inferred from schemas
export type LatLon = z.infer<typeof LatLonSchema>;
export type ShortestResponse = z.infer<typeof ShortestResponseSchema>;
export type MapNodesResponse = z.infer<typeof MapNodesResponseSchema>;
